src/lib/Imp.hs

```diff
@@ -454,544 +623,411 @@ splitDest (maybeDest, (Block decls ans)) = do
+
+variableScope :: ImpM Scope
+variableScope = looks envScope
+
+freshVar :: VarP a -> ImpM (VarP a)
+freshVar (hint:>t) = do
+  scope <- looks envScope
+  let v = genFresh (rawName GenName $ nameTag hint) scope
+  extend $ mempty { envScope = v @> (UnitTy, UnknownBinder) }
+  return $ v:>t
+
+withLevel :: ParallelismLevel -> ImpM a -> ImpM a
+withLevel level = local (\opts -> opts {curLevel = level })
+
+withDevice :: Device -> ImpM a -> ImpM a
+withDevice device = local (\opts -> opts {curDevice = device })
 
 -- === type checking imp programs ===
 
+-- TODO: track parallelism level and current device and add validity checks like
+-- "shouldn't launch a kernel from device/thread code"
+
 -- State keeps track of _all_ names used in the program, Reader keeps the type env.
-type ImpCheckM a = StateT (Env ()) (ReaderT (Env IType) (Either Err)) a
+type ImpCheckM a = StateT (Env ()) (ReaderT (Env IType, Device) (Either Err)) a
+
+instance Checkable ImpModule where
+  -- TODO: check main function defined
+  checkValid (ImpModule fs) = mapM_ checkValid fs
 
 instance Checkable ImpFunction where
-  checkValid (ImpFunction _ vsIn prog) = do
-    let scope = foldMap (binderAsEnv . fmap (const ())) vsIn
-    let env   = foldMap (binderAsEnv . fmap (IRefType . dropArray)) vsIn
-    void $ runReaderT (runStateT (checkProg prog) scope) env
-    where dropArray ~(ArrayTy t) = t
-
-checkProg :: IProg ImpInstr -> ImpCheckM ()
-checkProg prog = () <$ checkProgVal (prog, Nothing)
-
-checkProgVal :: IProgVal ImpInstr -> ImpCheckM (Maybe IType)
-checkProgVal (Empty, val) = traverse checkIExpr val
-checkProgVal ((Nest stmt prog), val) = do
-  env <- case stmt of
-    IInstr (binder, instr) -> do
-      ty <- instrTypeChecked instr
-      case (binder, ty) of
-        (IDo    , IVoidType) -> return mempty
-        (Bind _ , IVoidType) -> throw CompilerErr $ "Can't assign result of void instruction"
-        (b, _) -> do
-          checkBinder b
-          checkValidType $ binderAnn b
-          assertEq (binderAnn b) ty "Type mismatch in instruction"
-          return (b@>ty)
-    IFor _ i size block -> do
-      checkInt size
-      checkBinder i
-      assertEq (binderAnn i) (impExprType size) $ "Mismatch between the loop iterator and upper bound type"
-      extendR (i @> impExprType size) $ checkProg block
-      return mempty
-    IWhile cond body -> do
-      Just condTy <- checkProgVal cond
-      assertEq (IValType $ Scalar Int8Type) condTy $ "Not a bool: " ++ pprint cond
-      checkProg body
-      return mempty
-    ICond predicate consequent alternative -> do
-      predTy <- checkIExpr predicate
-      assertEq (IValType $ Scalar Int8Type) predTy "Type mismatch in predicate"
-      checkProg consequent
-      checkProg alternative
-      return mempty
-  extendR env $ checkProgVal (prog, val)
-
-instrTypeChecked :: ImpInstr -> ImpCheckM IType
+  checkValid f@(ImpFunction (_:> IFunType cc _ _) bs block) = addContext ctx $ do
```

Comment:
We should also check that no function arguments are pointers with the `Stack` address space, since we're not tagging stacks by the device